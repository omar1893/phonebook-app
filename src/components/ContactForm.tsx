import React, { useEffect, useState } from "react";
import { ContactFormProps } from "../types";
import { isInvalid } from "../utils/validation";
import { motion } from "framer-motion";

const ContactForm = ({
  data,
  setters,
  handleSubmit,
  editing,
}: ContactFormProps) => {
  const { firstName, lastName, phoneNumber } = data;
  const { setFirstName, setLastName, setPhoneNumber } = setters;
  const [isInvalidForm, setIsInvalidForm] = useState(true);

  useEffect(() => {
    setIsInvalidForm(isInvalid(firstName, lastName, phoneNumber));
  }, [firstName, lastName, phoneNumber]);

  const buttonVariants = {
    disabled: { opacity: 0.5, scale: 0.6 },
    enabled: { opacity: 1, scale: 1 },
  };

  return (
    <div className="block max-w-sm p-6 bg-white border border-gray-200 rounded-lg shadow dark:bg-gray-800 dark:border-gray-700 mx-auto">
      <form onSubmit={handleSubmit} className="flex flex-col">
        <div className="mb-6">
          <label
            htmlFor="base-input"
            className="block mb-2 text-sm font-medium text-gray-900 dark:text-white text-left"
          >
            Name
          </label>
          <input
            type="text"
            id="base-input"
            value={firstName}
            onChange={(e) => setFirstName(e.target.value)}
            className="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
          />
        </div>

        <div className="mb-6">
          <label
            htmlFor="base-input"
            className="block mb-2 text-sm font-medium text-gray-900 dark:text-white text-left"
          >
            Last Name
          </label>
          <input
            type="text"
            id="base-input"
            value={lastName}
            onChange={(e) => setLastName(e.target.value)}
            className="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
          />
        </div>

        <div className="mb-6">
          <label
            htmlFor="base-input"
            className="block mb-2 text-sm font-medium text-gray-900 dark:text-white text-left"
          >
            Phone Number
          </label>
          <input
            type="text"
            id="base-input"
            value={phoneNumber}
            onChange={(e) => setPhoneNumber(e.target.value)}
            className="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
          />
        </div>

        <motion.button
          type="submit"
          className="text-white bg-gradient-to-r from-green-400 via-green-500 to-green-600 hover:bg-gradient-to-br focus:ring-4 focus:outline-none focus:ring-green-300 dark:focus:ring-green-800 font-medium rounded-lg text-sm px-5 py-2.5 text-center mr-2 mb-2"
          variants={buttonVariants}
          animate={isInvalidForm ? "disabled" : "enabled"}
          disabled={isInvalidForm}
        >
          {editing ? "Edit" : "Add Contact"}
        </motion.button>
      </form>
    </div>
  );
};

export default ContactForm;
