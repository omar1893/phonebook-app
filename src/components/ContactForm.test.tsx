import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import ContactForm from "./ContactForm";

describe("ContactForm component", () => {
  test("should render the component correctly", () => {
    const data = { firstName: "", lastName: "", phoneNumber: "" };
    const setters = {
      setFirstName: jest.fn(),
      setLastName: jest.fn(),
      setPhoneNumber: jest.fn(),
    };
    const handleSubmit = jest.fn();
    const editing = false;

    const { container } = render(
      <ContactForm
        data={data}
        setters={setters}
        handleSubmit={handleSubmit}
        editing={editing}
      />
    );

    expect(container).toMatchSnapshot();
  });

  test("should disable the button when the form is invalid", () => {
    const data = { firstName: "", lastName: "", phoneNumber: "" };
    const setters = {
      setFirstName: jest.fn(),
      setLastName: jest.fn(),
      setPhoneNumber: jest.fn(),
    };
    const handleSubmit = jest.fn();
    const editing = false;

    render(
      <ContactForm
        data={data}
        setters={setters}
        handleSubmit={handleSubmit}
        editing={editing}
      />
    );

    const submitButton = screen.getByRole("button", { name: "Add Contact" }) as HTMLButtonElement;

    expect(submitButton.disabled).toBeTruthy();
  });

  test("should enable the button when the form is valid", () => {
    const data = { firstName: "John", lastName: "Doe", phoneNumber: "1234567890" };
    const setters = {
      setFirstName: jest.fn(),
      setLastName: jest.fn(),
      setPhoneNumber: jest.fn(),
    };
    const handleSubmit = jest.fn();
    const editing = false;

    render(
      <ContactForm
        data={data}
        setters={setters}
        handleSubmit={handleSubmit}
        editing={editing}
      />
    );

    const submitButton = screen.getByRole("button", { name: "Add Contact" }) as HTMLButtonElement;

    expect(submitButton.disabled).toBeFalsy();
  });

  test("should call handleSubmit when the form is submitted", () => {
    const data = { firstName: "John", lastName: "Doe", phoneNumber: "1234567890" };
    const setters = {
      setFirstName: jest.fn(),
      setLastName: jest.fn(),
      setPhoneNumber: jest.fn(),
    };
    const handleSubmit = jest.fn();
    const editing = false;

    render(
      <ContactForm
        data={data}
        setters={setters}
        handleSubmit={handleSubmit}
        editing={editing}
      />
    );

    const submitButton = screen.getByRole("button", { name: "Add Contact" });
    fireEvent.click(submitButton);

    expect(handleSubmit).toHaveBeenCalled();
  });
});