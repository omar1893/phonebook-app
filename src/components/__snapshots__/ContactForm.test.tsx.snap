// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm component should render the component correctly 1`] = `
<div>
  <div
    class="block max-w-sm p-6 bg-white border border-gray-200 rounded-lg shadow dark:bg-gray-800 dark:border-gray-700 mx-auto"
  >
    <form
      class="flex flex-col"
    >
      <div
        class="mb-6"
      >
        <label
          class="block mb-2 text-sm font-medium text-gray-900 dark:text-white text-left"
          for="base-input"
        >
          Name
        </label>
        <input
          class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
          id="base-input"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-6"
      >
        <label
          class="block mb-2 text-sm font-medium text-gray-900 dark:text-white text-left"
          for="base-input"
        >
          Last Name
        </label>
        <input
          class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
          id="base-input"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-6"
      >
        <label
          class="block mb-2 text-sm font-medium text-gray-900 dark:text-white text-left"
          for="base-input"
        >
          Phone Number
        </label>
        <input
          class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
          id="base-input"
          type="text"
          value=""
        />
      </div>
      <button
        class="text-white bg-gradient-to-r from-green-400 via-green-500 to-green-600 hover:bg-gradient-to-br focus:ring-4 focus:outline-none focus:ring-green-300 dark:focus:ring-green-800 font-medium rounded-lg text-sm px-5 py-2.5 text-center mr-2 mb-2"
        disabled=""
        type="submit"
      >
        Add Contact
      </button>
    </form>
  </div>
</div>
`;
